test_classes: Rename and refactor 'tornado_redirected_to_list'.

prakhar1144 opened the pull request:
This PR renames the `tornado_redirected_to_list` context manager to `capture_send_event_calls` to improve readability.

It also refactors the function to yield a list of events instead of passing in a list data structure as a parameter and appending events to it.

Previously:
```
with self.tornado_redirected_to_list(events, expected_num_events=1):
     do_something()
     ...
self.assertEqual(events[0], {...})
```

Now:
```
with self.capture_send_event_calls(expected_num_events=1) as events:
     do_something()
     ...
self.assertEqual(events[0], {...})
```

[**CZO Discussion**](https://chat.zulip.org/#narrow/stream/3-backend/topic/tornado_redirected_to_list/near/1540847)

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alexmv commented on zerver/lib/test_classes.py:
```diff
@@ -1692,10 +1692,10 @@ def subscribe_realm_to_monthly_plan_on_manual_license_management(
-    ) -> Iterator[None]:
-        lst.clear()
+    def capture_send_event_calls(
+        self, expected_num_events: int
```
I believe [the proposal](https://chat.zulip.org/#narrow/stream/3-backend/topic/tornado_redirected_to_list/near/1540990) was to rename this argument `expected_events`.

prakhar1144 replied:
After [discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/tornado_redirected_to_list/near/1541855), we are planning to keep using `expected_num_events`

alexmv commented on zerver/tests/test_bots.py:
```diff
@@ -164,7 +164,7 @@ def test_add_bot(self) -> None:
         self.login("hamlet")
         self.assert_num_bots_equal(0)
         events: List[Mapping[str, Any]] = []
```
All of these pre-definitions of `events` are now unnecessary -- that's one of the reasons to switch to the contextmanager returning the list.

prakhar1144 replied:
Makes sense! -- updated.

alexmv reviewed (changes requested):
Thanks for this cleanup!  I think we can do a little more polish by removing some now-unnecessary variable definitions, too.

I would perhaps contend that this doesn't "update" `tornado_redirected_to_list`, it renames and refactors it.  You can trade out "to improve readability" from the first line of the commit message (because that's pretty implied) and use the spare characters to say "rename and refactor" instead of "update".

prakhar1144 commented:
@alexmv This is ready for another look from you.

Updates in the recent push:
* Addressed the review comments. (I have replied to the respective review comments)

timabbott commented:
This is great, thanks so much for cleaning this up @prakhar1144!
